llvm/lib/Transforms/Yk/ControlPoint.cpp

```diff
@@ -0,0 +1,343 @@
+//===- ControlPoint.cpp - Synthesise the yk control point -----------------===//
+//
+// This pass finds the user's call to the dummy control point and replaces it
+// with a call to a new control point that implements the necessary logic to
+// drive the yk JIT.
+//
+// The pass converts an interpreter loop that looks like this:
+//
+// ```
+// pc = 0;
+// while (...) {
+//     yk_control_point() // <- dummy control point
+//     bc = program[pc];
+//     switch (bc) {
+//         // bytecode handlers here.
+//     }
+// }
+// ```
+//
+// Into one that looks like this (note that this transformation happens at the
+// IR level):
+//
+// ```
+// // The IO struct contains one member for each live LLVM variable at the
+// // control point.
+// struct IOStruct {
+//     size_t pc;
+// }
+//
+// pc = 0;
+// while (...) {
+//     struct IOStruct io = { pc };
+//     io = yk_new_control_point(io) // <- patched control point.
+//     pc = io.pc;
+//     bc = program[pc];
+//     switch (bc) {
+//         // bytecode handlers here.
+//     }
+// }
+// ```
+//
+// The call to the dummy control point must be the first thing that appears in
+// an interpreter dispatch loop.
+//
+// FIXME: The control point cannot yet be used in an interpreter using threaded
+// dispatch.
+//
+// FIXME: The tracing logic is currently over-simplified:
+//
+//  - A JIT location is assumed to be a simple integer program counter. It
+//    should be a `ykrt::Location`.
+//
+//  - Tracing starts when we encounter a location for which we have no machine
+//    code. A hot counter should be used instead.
+//
+//  - There can be only one compiled trace for now. There should be a code
+//    cache mapping from JIT locations to their machine code.
+//
+//  - The interpreter is assumed to be single threaded. We should implement a
```

Comment:
True though this is, I believe that *this* code doesn't have to worry about threading at all, and that this bullet point is probably irrelevant / misleading. All the threading stuff will be handled in Rust.